docs/docs/reference/metaprogramming/macros-spec.md

```diff
@@ -183,7 +183,7 @@ private def dynamicPower(n: Int, x: Double): Double =
   else if (n % 2 == 0) dynamicPower(n / 2, x * x)
   else x * dynamicPower(n - 1, x)
 ```
----
+
 This assumes a `Constant` extractor that maps tree nodes representing
 constants to their values.
```

Comment:
@nicolasstucki : do we still support `Constant`?

Reply:
@liufengyun For some reason the PNG file `explicit-nulls-type-hierarchy.png` is outside the `docs/reference/` directory (see [`images/explicit-nulls/`](https://github.com/lampepfl/dotty/tree/master/docs/images/explicit-nulls)). Before running `pandoc` I call a small preprocessor script which performs several transformations, e.g.
- replace internal relative links to anchors
- copy image files to the right place for `pandoc`

Reply:
Yes, but that code is extremely outdated and would not compile. I would use the following version instead

----------

```scala
inline def power(x: Double, n: Int): Double =
  ${ powerExpr('x, 'n) }

private def powerExpr(n: Expr[Int], x: Expr[Double])(using Quotes): Expr[Double] =
  n.value match {
    case Some(m) => powerCode(x, m)
    case _ => '{ dynamicPower($x, $n) }
  }

private def dynamicPower(x: Double, n: Int): Double =
  if (n == 0) 1.0
  else if (n % 2 == 0) dynamicPower(x * x, n / 2)
  else x * dynamicPower(x, n - 1)
```

This assumes a `.value` that maps tree nodes representing constants to their values.

----------

Reply:
@nicolasstucki Is the following self-contained code ok for you (NB. `powerCode` is not defined in your code) ?!

For replacement of the first code example in section "Going Further" of file [`metaprogramming/macro-spec.md`](https://github.com/lampepfl/dotty/blob/master/docs/docs/reference/metaprogramming/macros-spec.md):
```scala
import scala.quoted.Expr
import scala.quoted.Quotes

inline def power(x: Double, n: Int): Double =
  ${ powerExpr('x, 'n) }

private def powerExpr(x: Expr[Double], n: Expr[Int])(using Quotes): Expr[Double] =
  n.value match
    case Some(m) => powerExpr(x, m)
    case _ => '{ dynamicPower($x, $n) }

private def powerExpr(x: Expr[Double], n: Int)(using Quotes): Expr[Double] =
  if n == 0 then '{ 1.0 }
  else if n == 1 then x
  else if n % 2 == 0 then '{ val y = $x * $x; ${ powerExpr('y, n / 2) } }
  else '{ $x * ${ powerExpr(x, n - 1) } }

private def dynamicPower(x: Double, n: Int): Double =
  if n == 0 then 1.0
  else if n % 2 == 0 then dynamicPower(x * x, n / 2)
  else x * dynamicPower(x, n - 1)
```
For testing (not to be included): `Main.scala`
```scala
@main
def Main: Unit =
  val x = 2.0
  for n <- 1 to 10 do
    println(s"power($x, $n) = ${power(x, n)}")
```

**PS.** The paragraph directly following the above code example mentions `Constant`. I assume that text can be removed right ?!

Reply:
You can replace it with 
```markdown
This assumes a `.value` that maps tree nodes representing constants to their values.
```

Reply:
And use `import scala.quoted._`

Reply:
@liufengyun addressed review comment.